File: nervecenter/src/main/java/com/braindroid/nervecenter/playbackTools/ManagedMediaPlayerPool.java
package com.braindroid.nervecenter.playbackTools;

import com.braindroid.nervecenter.recordingTools.models.utils.PersistedRecordingFileHandler;

public class ManagedMediaPlayerPool {

    private final PersistedRecordingFileHandler fileHandler;

    private static class PooledMediaPlayer extends ManagedMediaPlayer {

        private ManagedMediaPlayerPool playerPool;

        private PooledMediaPlayer(PersistedRecordingFileHandler fileHandler, ManagedMediaPlayerPool playerPool) {
            super(fileHandler);
            this.playerPool = playerPool;
        }

        @Override
        public void play() {
            if(playerPool.currentMediaPlayer != null) {
                playerPool.currentMediaPlayer.pause();
            }
            playerPool.currentMediaPlayer = this;
            super.play();
        }
    }

    private ManagedMediaPlayer currentMediaPlayer = null;

    public ManagedMediaPlayerPool(PersistedRecordingFileHandler fileHandler) {
        this.fileHandler = fileHandler;
    }

    public ManagedMediaPlayer createManagedMediaPlayer() {
        return new PooledMediaPlayer(fileHandler, this);
    }

    public void stopPlayback() {
        if(currentMediaPlayer != null) {
            currentMediaPlayer.pause();
        }
    }
}

File: nervecenter/src/main/java/com/braindroid/nervecenter/kotlinModels/playbackTools/ManagedMediaPlayerPool.kt
package com.braindroid.nervecenter.kotlinModels.playbackTools

import com.braindroid.nervecenter.kotlinModels.utils.OnDiskRecordingFileHandler

class ManagedMediaPlayerPool(val diskRecordingFileHandler: OnDiskRecordingFileHandler) {

    class PooledPlayer(
            diskRecordingFileHandler: OnDiskRecordingFileHandler,
            val playerPool: ManagedMediaPlayerPool
    ) : ManagedMediaPlayer(diskRecordingFileHandler) {
        override fun play() {
            playerPool.currentPlayer?.let {
                if (it.isPlaying()) it.pause()
            }
            playerPool.currentPlayer = this
            super.play()
        }
    }

    var currentPlayer: ManagedMediaPlayer? = null

    fun fromPool() = PooledPlayer(diskRecordingFileHandler, this)
}